sqlliveness/slstorage,timeutil: various improvements

ajwerner opened the pull request:
See individual commits. This PR addresses some anxieties I had about the implementation of slstorage. In particular, it makes the testing deterministic by giving it a new table independent of the host cluster's state and allows injecting time. It then changes the GC to be less frequent and instead relies on failing out sessions when it determines they are expired. This is important to avoid spinning between when a session is deemed expired and when it would later be deleted by a gc interval. It also jitters that GC interval to avoid problems due to contention in larger clusters.

Commit:
slinstance: change default heartbeat interval

Once a second seems quite fast. Even our epoch-based leases aren't that fast
and they time out after 7 seconds.

Release note: None

Commit:
timeutil: move TimeSource and ManualTimeSource from quotapool

These are generally useful tools to control not just the clock but associated
timers. These tools are helpful when testing code that attempts to do something
with some periodicity.

Release note: None

Commit:
sqlliveness/slstorage: rework testing to be deterministic

The old testing facilities relied on interacting with the state of the
subsystem in a running cluster. It also relied on real clocks. This
made it difficult to write precise tests without serious hacks.

Release note: None

Commit:
sqlliveness/slstorage: increase gc interval, add mechanism to delete on fetch

Running the GC every is very extreme in order to be responsive to failures.
In general, if we observe that an expiration hasn't changed when we go to
fetch, we should just delete it then. This makes the GC interval much less
important for not a lot of cost. The GC process is not cheap and will contend
with updates.

Release note: None

Commit:
sqlliveness/slstorage: add jitter to GC

Periodic loops which contend for the same data can cause real problems if
they are aligned with system startup. Let's jitter the loop and not run it
immediately upon startup.

Release note: None

ajwerner reviewed (commented):
I wish I had set it up to review each commit separately, they are self-contained :(

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53196#-:-MFGPXlT5G1p0X5Ax9s6:b7yii2o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)



<!-- Sent from Reviewable.io -->

spaskob reviewed (approved):
> Reviewed 6 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)



<!-- Sent from Reviewable.io -->

Commit:
sqlliveness/slstorage: delete dead sessions from the live cache

This wasn't really a correctness problem as we already checked the dead
cache but if we determine a session is dead, why keep it around?

Also adds a TODO that was in my head.

Release note: None

ajwerner commented:
bors r+
